app/src/main/scala/org/alephium/explorer/service/TransactionHistoryService.scala

```diff
@@ -0,0 +1,233 @@
+                                      databaseConfig: DatabaseConfig[PostgresProfile],
+                                      scheduler: Scheduler,
+                                      gs: GroupSetting): Future[Unit] =
+    Future.successful(
+      scheduler.scheduleLoopAndForget(
+        taskId        = TransactionHistoryService.productPrefix,
+        firstInterval = ScalaDuration.Zero,
+        loopInterval  = interval
+      )(syncOnce())
+    )
+
+  def syncOnce()(implicit ec: ExecutionContext,
+                 dc: DatabaseConfig[PostgresProfile],
+                 gs: GroupSetting): Future[Unit] = {
+    logger.debug("Updating transactions count")
+    val startedAt = TimeStamp.now()
+    updateTxHistoryCount().map { _ =>
+      val duration = TimeStamp.now().deltaUnsafe(startedAt)
+      logger.debug(s"Transactions history updated in ${duration.millis} ms")
+    }
+  }
+
+  def getPerChain(from: TimeStamp, to: TimeStamp, intervalType: IntervalType)(
+      implicit ec: ExecutionContext,
+      dc: DatabaseConfig[PostgresProfile]): Future[Seq[TimedValues]] = {
+    run(getPerChainQuery(intervalType, from, to)).map(_.groupBy {
+      case (timestamp, _, _, _) => timestamp
+    }.map {
+      case (timestamp, values) =>
+        val total = values.map { case (_, _, _, count) => count }.sum
+        val perChainValues = values.map {
+          case (_, chainFrom, chainTo, count) =>
+            PerChainValue(chainFrom.value, chainTo.value, count)
+        }
+        TimedValues(timestamp, total, Some(perChainValues))
+    }.toSeq)
+  }
+
+  def getAllChains(from: TimeStamp, to: TimeStamp, intervalType: IntervalType)(
+      implicit ec: ExecutionContext,
+      dc: DatabaseConfig[PostgresProfile]): Future[Seq[(TimeStamp, Long)]] = {
+    run(
+      getAllChainsQuery(intervalType, from, to)
+    )
+  }
+
+  private def updateTxHistoryCount()(implicit ec: ExecutionContext,
+                                     dc: DatabaseConfig[PostgresProfile],
+                                     gs: GroupSetting): Future[Unit] = {
+    findLatestTransationTimestamp().flatMap {
+      case None => Future.successful(()) //noop
+      case Some(latestTxTs) =>
+        for {
+          _ <- updateTxHistoryCountForInterval(latestTxTs, IntervalType.Daily)
+          _ <- updateTxHistoryCountForInterval(latestTxTs, IntervalType.Hourly)
+        } yield ()
+    }
+  }
+
+  def updateTxHistoryCountForInterval(latestTxTs: TimeStamp, intervalType: IntervalType)(
+      implicit ec: ExecutionContext,
+      dc: DatabaseConfig[PostgresProfile],
+      gs: GroupSetting): Future[Unit] = {
+    run(findLatestHistoryTimestamp(intervalType)).flatMap { histTsOpt =>
+      val start = histTsOpt
+        .map { histTs =>
+          stepBack(histTs, intervalType)
+        }
+        .getOrElse(ALPH.LaunchTimestamp)
+
+      val ranges = getTimeRanges(start, latestTxTs, intervalType)
+
+      foldFutures(ranges) {
```

Comment:
For more parallelism do we wanna consider using `Future.traverse(ranges)`? Assuming `ranges` contains dates that are not overlapping.

Reply:
actually when you sync the first time or you didn't sync for several days, it fails with a `traverse` because the db can't handle that much connections and inserts at the same time.

Reply:
Ah yep! Makes sense! 👍🏼  This probably should be looked into by #204: motivation being that we shouldn't have to reduce in-code performance (CPU based operations) to satisfy performance deficiency is database or any IO, this should be handled within thread-pools and connection-pools.

Reply:
Fully agree :+1: